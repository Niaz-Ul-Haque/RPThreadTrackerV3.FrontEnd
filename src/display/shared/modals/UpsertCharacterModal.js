import React from 'react';
import PropTypes from 'prop-types';
import { Modal, ModalHeader, ModalBody, ModalFooter, Button } from 'reactstrap';
import { AvForm } from 'availity-reactstrap-validation';
import UpsertCharacterForm from '../../forms/upsert-character/UpsertCharacterForm';
import TooltipForm from '../../forms/TooltipForm';

const propTypes = {
	isUpsertCharacterModalOpen: PropTypes.bool.isRequired,
	submitUpsertCharacter: PropTypes.func.isRequired,
	closeUpsertCharacterModal: PropTypes.func.isRequired,
	characterToEdit: PropTypes.shape({}).isRequired
};

class UpsertCharacterModal extends React.Component {
	constructor(props) {
		super(props);
		this.handleInputChange = this.handleInputChange.bind(this);
		this.state = {
			characterToEdit: props.characterToEdit
		};
	}

	componentWillReceiveProps(nextProps) {
		this.setState({ characterToEdit: nextProps.characterToEdit });
	}

	handleInputChange(event) {
		const { target } = event;
		const value = target.type === 'checkbox' ? target.checked : target.value;
		const { name } = target;
		this.setState({
			characterToEdit: Object.assign({}, this.state.characterToEdit, {
				[name]: value
			})
		});
	}
	render() {
		const {
			isUpsertCharacterModalOpen,
			submitUpsertCharacter,
			closeUpsertCharacterModal,
			characterToEdit
		} = this.props;
		if (!characterToEdit) {
			return (
				<div />
			);
		}
		return (
			<Modal isOpen={isUpsertCharacterModalOpen} toggle={closeUpsertCharacterModal} backdrop>
				<AvForm onValidSubmit={() => submitUpsertCharacter(this.state.characterToEdit)}>
					<ModalHeader toggle={closeUpsertCharacterModal}>{characterToEdit.characterId ? 'Edit Character' : 'Add Character'}</ModalHeader>
					<ModalBody>
						<TooltipForm
							Renderable={UpsertCharacterForm}
							characterToEdit={characterToEdit}
							handleInputChange={this.handleInputChange}
						/>
					</ModalBody>
					<ModalFooter>
						<Button color="primary">Submit Character</Button>{' '}
						<Button color="secondary" onClick={closeUpsertCharacterModal}>Cancel</Button>
					</ModalFooter>
				</AvForm>
			</Modal>
		);
	}
}

UpsertCharacterModal.propTypes = propTypes;
export default UpsertCharacterModal;
