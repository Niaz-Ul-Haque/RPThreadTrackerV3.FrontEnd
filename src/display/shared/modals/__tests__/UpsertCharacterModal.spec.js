// #region imports
import React from 'react';
import { shallow } from 'enzyme';
import { getSpecWrapper } from '../../../../../config/tests/helpers.unit';
import UpsertCharacterModal from '../UpsertCharacterModal';
// #endregion imports

// #region mocks
jest.mock('../../../forms/upsert-character/UpsertCharacterForm', () => 'UpsertCharacterForm');
jest.mock('../../../forms/TooltipForm', () => 'TooltipForm');
// #endregion mocks

const createTestProps = propOverrides => ({
	isUpsertCharacterModalOpen: true,
	submitUpsertCharacter: jest.fn(),
	closeUpsertCharacterModal: jest.fn(),
	characterToEdit: { characterId: 1 },
	...propOverrides
});

describe('rendering', () => {
	describe('snapshots', () => {
		it('should render valid snapshot', () => {
			const props = createTestProps();
			const jsx = (<UpsertCharacterModal {...props} />);
			const element = shallow(jsx);
			expect(element).toMatchSnapshot();
		});
		it('should render valid snapshot if character is new', () => {
			const props = createTestProps({ characterToEdit: {} });
			const jsx = (<UpsertCharacterModal {...props} />);
			const element = shallow(jsx);
			expect(element).toMatchSnapshot();
		});
	});
});

describe('behavior', () => {
	describe('submitUpsertCharacter', () => {
		it('should be called when form is submitted', () => {
			const submitUpsertCharacter = jest.fn();
			const characterToEdit = { characterId: 2 };
			const props = createTestProps({ submitUpsertCharacter, characterToEdit });
			const jsx = (<UpsertCharacterModal {...props} />);
			const element = shallow(jsx);
			const form = getSpecWrapper(element, 'upsert-character-modal-form');
			form.prop('onValidSubmit')();
			expect(submitUpsertCharacter).toHaveBeenCalledTimes(1);
			expect(submitUpsertCharacter).toHaveBeenLastCalledWith(characterToEdit);
		});
	});
	describe('handleInputChange', () => {
		it('should handle text field update', () => {
			const submitUpsertCharacter = jest.fn();
			const event = { target: { type: 'text', name: 'characterName', value: 'My Character' } };
			const props = createTestProps({ submitUpsertCharacter });
			const jsx = (<UpsertCharacterModal {...props} />);
			const element = shallow(jsx);
			element.instance().handleInputChange(event);
			element.update();
			const form = getSpecWrapper(element, 'upsert-character-modal-form');
			form.prop('onValidSubmit')();
			expect(submitUpsertCharacter).toHaveBeenCalledTimes(1);
			expect(submitUpsertCharacter).toHaveBeenLastCalledWith({ characterId: 1, characterName: 'My Character' });
		});
		it('should handle checkbox field update', () => {
			const submitUpsertCharacter = jest.fn();
			const event = { target: { type: 'checkbox', name: 'isActive', checked: true } };
			const props = createTestProps({ submitUpsertCharacter });
			const jsx = (<UpsertCharacterModal {...props} />);
			const element = shallow(jsx);
			element.instance().handleInputChange(event);
			element.update();
			const form = getSpecWrapper(element, 'upsert-character-modal-form');
			form.prop('onValidSubmit')();
			expect(submitUpsertCharacter).toHaveBeenCalledTimes(1);
			expect(submitUpsertCharacter).toHaveBeenLastCalledWith({ characterId: 1, isActive: true });
		});
	});
	describe('componentWillReceiveProps', () => {
		it('should set character to edit', () => {
			const submitUpsertCharacter = jest.fn();
			const characterToEdit = { characterId: 2 };
			const props = createTestProps({ submitUpsertCharacter });
			const jsx = (<UpsertCharacterModal {...props} />);
			const element = shallow(jsx);
			element.setProps({ characterToEdit });
			const form = getSpecWrapper(element, 'upsert-character-modal-form');
			form.prop('onValidSubmit')();
			expect(submitUpsertCharacter).toHaveBeenCalledTimes(1);
			expect(submitUpsertCharacter).toHaveBeenLastCalledWith(characterToEdit);
		});
	});
	describe('closeUpsertCharacterModal', () => {
		it('should be triggered when modal is toggled', () => {
			const closeUpsertCharacterModal = jest.fn();
			const props = createTestProps({ closeUpsertCharacterModal });
			const jsx = (<UpsertCharacterModal {...props} />);
			const element = shallow(jsx);
			const modal = getSpecWrapper(element, 'upsert-character-modal');
			modal.prop('toggle')();
			expect(closeUpsertCharacterModal).toHaveBeenCalledTimes(1);
		});
		it('should be triggered when modal header is toggled', () => {
			const closeUpsertCharacterModal = jest.fn();
			const props = createTestProps({ closeUpsertCharacterModal });
			const jsx = (<UpsertCharacterModal {...props} />);
			const element = shallow(jsx);
			const header = getSpecWrapper(element, 'upsert-character-modal-header');
			header.prop('toggle')();
			expect(closeUpsertCharacterModal).toHaveBeenCalledTimes(1);
		});
		it('should be triggered when close button is clicked', () => {
			const closeUpsertCharacterModal = jest.fn();
			const props = createTestProps({ closeUpsertCharacterModal });
			const jsx = (<UpsertCharacterModal {...props} />);
			const element = shallow(jsx);
			const button = getSpecWrapper(element, 'upsert-character-modal-close-button');
			button.simulate('click');
			expect(closeUpsertCharacterModal).toHaveBeenCalledTimes(1);
		});
	});
});
