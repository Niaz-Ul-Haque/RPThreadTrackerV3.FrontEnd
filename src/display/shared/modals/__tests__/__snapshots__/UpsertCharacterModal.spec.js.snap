// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering snapshots should render valid snapshot 1`] = `
<Styled(Modal)
  backdrop={true}
  data-spec="upsert-character-modal"
  isOpen={true}
  toggle={[MockFunction]}
>
  <AvForm
    data-spec="upsert-character-modal-form"
    method="get"
    model={Object {}}
    onInvalidSubmit={[Function]}
    onKeyDown={[Function]}
    onSubmit={[Function]}
    onValidSubmit={[Function]}
    tag={[Function]}
    validationEvent={
      Array [
        "onChange",
        "onInput",
      ]
    }
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      data-spec="upsert-character-modal-header"
      tag="h5"
      toggle={[MockFunction]}
      wrapTag="div"
    >
      Edit Character
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <TooltipForm
        Renderable={[Function]}
        characterToEdit={
          Object {
            "characterId": 1,
          }
        }
        handleInputChange={[Function]}
      />
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="primary"
        tag="button"
      >
        Submit Character
      </Button>
       
      <Button
        color="secondary"
        data-spec="upsert-character-modal-close-button"
        onClick={[MockFunction]}
        tag="button"
      >
        Cancel
      </Button>
    </ModalFooter>
  </AvForm>
</Styled(Modal)>
`;

exports[`rendering snapshots should render valid snapshot if character is new 1`] = `
<Styled(Modal)
  backdrop={true}
  data-spec="upsert-character-modal"
  isOpen={true}
  toggle={[MockFunction]}
>
  <AvForm
    data-spec="upsert-character-modal-form"
    method="get"
    model={Object {}}
    onInvalidSubmit={[Function]}
    onKeyDown={[Function]}
    onSubmit={[Function]}
    onValidSubmit={[Function]}
    tag={[Function]}
    validationEvent={
      Array [
        "onChange",
        "onInput",
      ]
    }
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      data-spec="upsert-character-modal-header"
      tag="h5"
      toggle={[MockFunction]}
      wrapTag="div"
    >
      Add Character
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <TooltipForm
        Renderable={[Function]}
        characterToEdit={Object {}}
        handleInputChange={[Function]}
      />
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="primary"
        tag="button"
      >
        Submit Character
      </Button>
       
      <Button
        color="secondary"
        data-spec="upsert-character-modal-close-button"
        onClick={[MockFunction]}
        tag="button"
      >
        Cancel
      </Button>
    </ModalFooter>
  </AvForm>
</Styled(Modal)>
`;
